src/dataflow/src/source/mod.rs

```diff
@@ -1361,23 +1375,32 @@ where
                     //
                     // Side note: with persistence enabled for the source, we will anyways never
                     // re-emit old data, because we start reading from the persisted offsets.
-                    let current_binding = timestamp_histories
-                        .get_binding(&source_ts.partition, source_ts.offset)
-                        .map(|(binding, _offset)| binding)
-                        .unwrap_or(0);
-                    if current_binding < assigned_ts.0 {
-                        timestamp_histories.add_binding(
-                            source_ts.partition.clone(),
-                            assigned_ts.0,
-                            source_ts.offset,
-                            false,
-                        );
-                    } else {
-                        tracing::debug!(
-                            "Filtered out timestamp binding {:?} from persistence because we already have {}.",
-                            (source_ts, assigned_ts),
-                            current_binding
-                        );
+                    let lower = Antichain::from_elem(assigned_ts.0);
+                    let upper = Antichain::new();
+                    // NOTE: We use get_bindings_in_range() and not the seemingly better
+                    // get_binding(). However, get_binding() does not only consult the actual
+                    // stored bindings but also checks the Proposer, and returns its output as a
+                    // binding if it doesn't have a stored binding. We only care about stored
+                    // bindings here.
+                    let existing_binding = timestamp_histories
+                        .get_bindings_in_range(lower.borrow(), upper.borrow()).into_iter()
+                        .filter(|(pid, _ts, _offset)| *pid == source_ts.partition).next();
```

Comment:
what if the binding in timestamp_histories actually corresponds to a lower offset, would we still want to keep it instead of the binding from persist? or is that not possible?

Reply:
ping @aljoscha

Reply:
That would be an invalid state. I don't think it can happen, though, but I will add a filter condition.